src/clj/puppetlabs/http/client/async.clj

```diff
@@ -117,11 +124,68 @@
       (.setAs (clojure-response-body-type->java opts))
       (.setBody (:body opts))
       (.setDecompressBody (clojure.core/get opts :decompress-body true))
-      (.setHeaders (:headers opts))))
+      (.setHeaders (:headers opts))
+      (.setMetricId (parse-metric-id opts))))
+
+(schema/defn get-mean :- schema/Num
+  [timer :- Timer]
+  (->> timer
+       .getSnapshot
+       .getMean
+       (.toMillis TimeUnit/NANOSECONDS)))
+
+(defn get-metric-data
+  [timer metric-id]
+  (let [count (.getCount timer)
+        mean (get-mean timer)
+        aggregate (* count mean)]
+    {:count count
+     :mean mean
+     :aggregate aggregate
+     :metric-id metric-id}))
+
+(defn get-metrics-data
+  [timers]
+  (reduce (fn [acc [metric-id timer]]
+            (assoc acc metric-id (get-metric-data timer metric-id)))
+          {} timers))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;; Public
 
+(schema/defn ^:always-validate get-client-metrics :- (schema/maybe common/Metrics)
+  "Returns the http client-specific metrics from the metric registry."
+  ([metric-registry :- common/OptionalMetricRegistry]
+   (when metric-registry
+     (into {} (Metrics/getClientMetrics metric-registry))))
+  ([metric-registry :- common/OptionalMetricRegistry
+    metric-filter :- common/MetricFilter]
+   (when metric-registry
+     (cond
+       (:verb metric-filter) (into {} (Metrics/getClientMetricsWithUrlAndVerb
+                                       metric-registry
+                                       (:url metric-filter)
+                                       (:verb metric-filter)
+                                       Metrics$MetricType/BYTES_READ))
+       (:url metric-filter) (into {} (Metrics/getClientMetricsWithUrl
+                                      metric-registry
+                                      (:url metric-filter)
+                                      Metrics$MetricType/BYTES_READ))
+       (:metric-id metric-filter) (into {} (Metrics/getClientMetricsWithMetricId
+                                            metric-registry
+                                            (into-array (:metric-id metric-filter)) Metrics$MetricType/BYTES_READ))
+       :else (throw (IllegalArgumentException. "Not an allowed metric filter."))))))
+
+(schema/defn ^:always-validate get-client-metrics-data :- (schema/maybe common/MetricsData)
+  "Returns a map of metric-id to metric data summary."
+  ([metric-registry :- common/OptionalMetricRegistry]
+   (let [timers (get-client-metrics metric-registry)]
+     (get-metrics-data timers)))
+  ([metric-registry :- common/OptionalMetricRegistry
+    metric-filter :- common/MetricFilter]
```

Comment:
>1. I think we may want to monkey with some of the names, e.g. `with-metric-id`, `bytes-read`, etc., but I feel like it should be pretty easy to put that off as like the last thing we do in this PR (or maybe even a follow-on PR).  So I'd vote for maybe belaying that conversation.

Yeah, that's all easy to change and I figured we'd probably want to change it over time.

>2. Constructing one of these `MetricFilter` schema objects by hand feels a bit cumbersome (only talking about Clojure here, I don't necessarily care whether the Java API is a bit cumbersome, because, you know, it's Java).  I think it might be worth creating some helper functions to help build these up.  Maybe, at the least, a function that you pass in a partial map and it fills in the rest with some default values?  Or, to get a bit more granular, maybe something like `(metric-id-filter ["puppetlabs"])` and `(url-filter "https://tickets.puppetlabs.com")`.

Sure? I guess I don't feel like it's that cumbersome/makes that much of a difference - `{:metric-id [:foo] :metric-type :bytes-read}` vs `(metric-id-filter [:foo] :bytes-read)`, `{:url "http://tickets.puppetlabs.com :metric-type :bytes-read}` vs `(url-filter "https://tickets.puppetlabs.com" :bytes-read)`, and` {:url "http://tickets.puppetlabs.com :method "GET" :metric-type :bytes-read}` vs `(url-fitler "https://tickets.puppetlabs.com" "GET" :bytes-read)`? Or we have `(bytes-read-filter ...)` and `(response-init-filter ...)` that handle adding in the metric-type? This seems like something that can be added in later without being a breaking change, so long as we keep the schema the same.

One thing I realize from this though is that maybe we should standardize the method name schema should be keywords :get, :post, etc, and then we convert that to what the method will show up as in the metric name.

>3. I notice that if I make a request to `https://confluence.puppetlabs.com` and then another one to `https://confluence.puppetlabs.com/` (with a trailing slash), these show up as two separate metrics on the `with-url` side of the house.  I think that's probably fine, and that if people wanted to aggregate those somehow they can always use a `metric-id`.  Just thought it was worth mentioning; it also made me wonder if we needed to think at all about special characters that might show up in a URL and whether there are any possible issues with how those might get mapped to a metric id.

It seems maybe reasonable to strip trailing slashes... but then do we also strip trailing question marks? What else should we strip? What about things that can optionally percent encoded... should we standardize those before making the metric id? Lollll terrible questions. Okay, I think based on going down this train of thought I am against standardizing the url in any way before making it a part of a metric name - that's why we have the metric id feature.

>5. I notice that query strings are included in the `with-url` metric ids.  Mixed feelings on this; I can see how they might be useful, but I can also imagine that leading to a large proliferation of timers, and make it hard or impossible to aggregate the URL metrics for a given URL ignoring the differences in the query string.  So, I could imagine us making the call that we just truncate query strings from the `with-url` metric ids, or I could imagine us maybe saying that in a future update we might add another type of `MetricFilter` that supports a regex value for the `:url` field.  Not sure.

I actually didn't realize that query strings were included in this. I think that they should be stripped out - I agree that they'll make for too many too specific timers, making it hard to aggregate.

>6. I think it would be useful to have a `MetricFilter` that allowed me to ask for all of the `with-metric-id` metrics, without having to also specify a specific `metric-id`.

Sure.


>7. When I make two requests, one with a metric-id of ["puppetlabs" "jira"] and one with a metric-id of ["puppetlabs" "confluence"], and then I call `get-client-metrics-data` with a filter that includes `:metric-id ["puppetlabs"]`, I *only* get back the aggregate data.  I think I was kind of expecting that I'd get that *and* the values for the child nodes in that hierarchy.  I can imagine both use cases being desirable, so maybe we could add something to the `MetricFilter` that would allow you to flag whether you wanted only the exact match or everything underneath that in the tree.  It's possible that we already discussed this in our design meeting and agreed on the current approach, I can't remember.

I think I had asked about this in hipchat, and I think in that discussion we had decided to do just the exact match. But I think a reasonable API would be something like ["puppetlabs" "*"] if you want puppetlabs and everything under it, and ["puppetlabs"] if you just want the puppetlabs metric